### PR TITLE
Enable calling recursive functions from the GPU

### PR DESCRIPTION
**Pull request by daviditen**
When the second instance of a function is detected in a call chain (meaning it is recursive) return that it is gpuizable. This will enable calling it from the GPU as long as the rest of the first instance encountered is OK.

The presence of recursive functions makes ptxas unable to statically determine the stack size, which it warns about by default. Add a flag "--suppress-stack-size-warning" to the list of flags for ptxas.

Add a test that recursively computes Fibonacci numbers on the GPU.

Update a test that had an infinite recursion to stop after a smaller number of calls.

### CODE REVIEW

File: compiler/llvm/clangUtil.cpp [hunk 1]
```diff
@@ -4144,6 +4144,9 @@ static void makeBinaryLLVMForCUDA(const std::string& artifactFilename,
     }
   }
 
+  // avoid warning about not statically knowing the stack size when recursive
+  // functions are called from the kernel
+  ptxasFlags += " --suppress-stack-size-warning ";
```
Comment (stonea) [inline compiler/llvm/clangUtil.cpp:4149]: Do we need to add a similar flag for AMD?

File: test/gpu/native/assertOnFailToGpuize.chpl [hunk 2]
```diff
@@ -22,14 +22,14 @@ on here.gpus[0] {
   if failureMode == 1 {
```
Comment (stonea) [inline test/gpu/native/assertOnFailToGpuize.chpl:22]: I think I'd remove the individual `.good` files for these cases that now pass and move them to the bottom of this file where we have a bunch of other loops that are expected to work. Then renumber the failure modes (e.g. move `assertOnFailToGpuize.3.good` through `assertOnFailToGpuize.5.good` to `assertOnFailToGpuize.1.good` to `assertOnFailToGpuize.3.good`, get rid of modes 6&7 and then have 6 be the "cases we expect to pass" mode, and then update the `.compopt` file to iterate from 1 to 6.)

It's kind of a pain, but I also kind of like having this test as kind of a self-documenting list of the known failure modes that `assertOnGpu()` can capture, and it's kind of misleading to have something that's meant to pass under `if failureMode == x { ... }`.

**Review by stonea — COMMENTED** [review]
Very cool, nice to see a limitation we've had for a while get removed. So we don't forget to do it before our next release, could you also update this line in the tech note to no longer mention recursive functions:

"They are free of any call to a function that fails to meet the above criteria, accesses outer variables, or are recursive"

Thanks!

File: compiler/llvm/clangUtil.cpp (hunk 1 above)
Comment (daviditen) [reply compiler/llvm/clangUtil.cpp:4149]: We don't seem to need it for AMD.

File: test/gpu/native/assertOnFailToGpuize.chpl (hunk 2 above)
Comment (daviditen) [reply test/gpu/native/assertOnFailToGpuize.chpl:22]: I updated the test do do this.

**Review by stonea — APPROVED** [review]
With https://github.com/chapel-lang/chapel/pull/22002/commits/b8b9a2187389f4d17c528919d26dbc46b3d9d36b fixing the AMD linking issue this looks good to go.

You might want to do a full `start_test` under `tests/gpu/native` with AMD just to make sure that doesn't inadvertently break anything (but I've got my fingers crossed).

**Comment by daviditen** [comment]
> With [b8b9a21](https://github.com/chapel-lang/chapel/commit/b8b9a2187389f4d17c528919d26dbc46b3d9d36b) fixing the AMD linking issue this looks good to go.
> 
> You might want to do a full `start_test` under `tests/gpu/native` with AMD just to make sure that doesn't inadvertently break anything (but I've got my fingers crossed).

[Summary: #Successes = 51 | #Failures = 0 | #Futures = 0 | #Warnings = 0 ]

**Pull request merged**